### PR TITLE
Update getValidDocIdsMetadataFromServer to make call in batches to servers and other bug fixes

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
label:
`bugfix`
`upsert`

This patch addresses multiple things:

1. We saw issue with upsert-compaction not working for few servers of a table but working for others. This was primarily because we send all segment names from controller to server in payload and even if one is missing (which can be due to concurrent deletion) it results in `NOT_FOUND` error. We are moving the failure behaviour of API in this scenario to a warn log instead and continue proceeding to return valid doc ids for whatever segments are available. 
2. We send high number of segments in the payload from controller to server. Sending requests in batches so that request-body is nominal in size and if there is issue with one segment, then entire host data is not missed but only batch's data.
3. For CompletionServiceHelper code, if we enabled `multiRequestsPerServer` flag and the uri is same for the requests to a server, then only last response will be returned back to the caller. We add a count value to the uri key in this scenario to honour all responses from a single host.

cc @klsince @snleee

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 1]
```diff
@@ -61,6 +61,7 @@
  */
 public class ServerSegmentMetadataReader {
   private static final Logger LOGGER = LoggerFactory.getLogger(ServerSegmentMetadataReader.class);
+  private static final int DEFAULT_SEGMENT_BATCH_SIZE_FOR_QUERYING_SERVER = 500;
```
Comment (tibrewalpratik17) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:64]: Discussed offline with @klsince , we want to expose this as a configurable param in upsert-compaction-task and in the controller GET call API as well.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 2]
```diff
@@ -972,7 +972,10 @@ public String getTableAggregateValidDocIdsMetadata(
       @ApiParam(value = "A list of segments", allowMultiple = true) @QueryParam("segmentNames")
       List<String> segmentNames,
       @ApiParam(value = "Valid doc ids type") @QueryParam("validDocIdsType")
-      @DefaultValue("SNAPSHOT") ValidDocIdsType validDocIdsType, @Context HttpHeaders headers) {
+      @DefaultValue("SNAPSHOT") ValidDocIdsType validDocIdsType,
+      @ApiParam(value = "Number of segments in a batch per server request")
+      @QueryParam("numSegmentsBatchPerServerRequest") @DefaultValue("500") int numSegmentsBatchPerServerRequest,
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:977]: nit: numSegmentsPerServerRequest or serverRequestBatchSize

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java [hunk 3]
```diff
@@ -657,9 +657,7 @@ private List<Map<String, Object>> processValidDocIdsMetadata(String tableNameWit
     }
     try {
       if (!missingSegments.isEmpty()) {
-        throw new WebApplicationException(
-            String.format("Table %s has missing segments: %s)", tableNameWithType, segments),
-            Response.Status.NOT_FOUND);
+        LOGGER.warn("Table {} has missing segments {}", tableNameWithType, missingSegments);
```
Comment (klsince) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:660]: can add a comment on why no need to abort

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java [hunk 4]
```diff
@@ -58,6 +58,7 @@ public class UpsertCompactionTaskGenerator extends BaseTaskGenerator {
   private static final String DEFAULT_BUFFER_PERIOD = "7d";
   private static final double DEFAULT_INVALID_RECORDS_THRESHOLD_PERCENT = 0.0;
   private static final long DEFAULT_INVALID_RECORDS_THRESHOLD_COUNT = 0;
+  private static final int DEFAULT_SEGMENT_BATCH_SIZE_FOR_QUERYING_SERVER = 500;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java:61]: nit: simply `DEFAULT_`NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 5]
```diff
@@ -235,8 +236,18 @@ public List<ValidDocIdsMetadataInfo> getValidDocIdsMetadataFromServer(String tab
           }
         }
       }
-      serverURLsAndBodies.add(generateValidDocIdsMetadataURL(tableNameWithType, segmentsToQuery, validDocIdsType,
-          serverToEndpoints.get(serverToSegments.getKey())));
+
+      // Number of segments to query per server request. If a table has a lot of segments, then we might send a
+      // huge payload to pinot-server in request. Batching the requests will help in reducing the payload size.
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:241]: use Lists.partition() for short?

**Review by klsince — APPROVED** [review]
LGTM, just some comments on variable naming

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/CompletionServiceHelper.java [hunk 6]
```diff
@@ -129,13 +129,21 @@ private CompletionServiceResponse collectResponse(String tableNameWithType, int
...
         String responseString = EntityUtils.toString(multiHttpRequestResponse.getResponse().getEntity());
-        completionServiceResponse._httpResponses
-            .put(multiRequestPerServer ? uri.toString() : instance, responseString);
+        String key = multiRequestPerServer ? uri.toString() : instance;
+        // If there are multiple requests to the same server with the same URI but different payloads,
+        // we append a count value to the key to ensure each response is uniquely identified.
+        // Otherwise, the map will store only the last response, overwriting previous ones.
+        if (multiRequestPerServer) {
+          int count = completionServiceResponse._instanceToRequestCount.getOrDefault(key, 0) + 1;
+          completionServiceResponse._instanceToRequestCount.put(key, count);
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/CompletionServiceHelper.java:143]: can do `compute(key, (k, v) -> (v == null) ? 1 : v + 1);` but not a blocking comment

**Review by klsince — APPROVED** [review]

**Pull request merged**